### PR TITLE
Add various Automaton adapters

### LINKED ISSUE #2: Fuzzy prefix searches?
Wow, what a slick library. I was just evaluating this for use in building custom indices, and I'm really impressed. I think I'm going to get a lot of mileage out of this. :-)

However, I ran into one use case that would be very handy: Fuzzy prefix matching. Given a string `Masach`, I can search for `Levenshtein::new("Masach", 1)` or `Regex::new("Masach.*")`. But what I'd really love to do is autocomplete to `Massachusetts`, with the extra `s`.

It seems like it should be possible to build something like:

```rust
// Match anything with a prefix that would be matched by A.
impl<A: Automaton> Automaton for Prefix<A> { ... }
```

But I'm not quite sure how to go about something like that. Is this basically trivial, or is it much harder than it looks?

### PR DESCRIPTION
**Pull request by gereeter**
This wants tests and better documentation, but I thought I'd get your thoughts on this before doing those. This fixes #2.

### CODE REVIEW

File: src/automaton.rs [hunk 1]
```diff
@@ -41,10 +43,49 @@ pub trait Automaton {
...
+    /// steps are taken.
+    ///
+    /// If this returns `true`, then every sequence of inputs from this state
+    /// produces a match. If this does not follow, then those match states may
+    /// never be reached. In other words, behavior may be incorrect.
+    ///
+    /// If this returns `false` even when every sequence of inputs will lead to
+    /// a match, then behavior will be correct, but callers may be forced to do
+    /// additional work.
+    fn must_match(&self, _state: &Self::State) -> bool {
```
Comment (BurntSushi) [inline src/automaton.rs]: Hmm, I feel like a better name for this method would be `will_always_match` or something. `must_match` is crossing some wires in my brain.

File: src/automaton.rs [hunk 2]
```diff
@@ -79,5 +124,171 @@ impl Automaton for AlwaysMatch {
...
+    } 
+}
+
+/// An automaton that matches when one of its component automata match.
+pub struct Union<A, B>(A, B);
+
+/// The `Automaton` state for `Union<A, B>`.
+pub struct UnionState<A: Automaton, B: Automaton>(A::State, B::State);
+
+impl<A: Automaton, B: Automaton> Automaton for Union<A, B> {
```
Comment (BurntSushi) [inline src/automaton.rs]: This impl is gorgeous.

File: src/automaton.rs [hunk 3]
```diff
@@ -79,5 +124,171 @@ impl Automaton for AlwaysMatch {
...
+    }
+}
+
+/// An automaton that matches when both of its component automata match.
+pub struct Intersection<A, B>(A, B);
+
+/// The `Automaton` state for `Intersection<A, B>`.
+pub struct IntersectionState<A: Automaton, B: Automaton>(A::State, B::State);
+
+impl<A: Automaton, B: Automaton> Automaton for Intersection<A, B> {
```
Comment (BurntSushi) [inline src/automaton.rs]: So is this. :-)

**Comment by BurntSushi** [comment]
Umm. YES PLEASE! :-) This looks fantastic. I had one little nit about the name, but I'd be happy to take this in unless you want to write a couple tests first. :-)

**Comment by gereeter** [comment]
Okay, name changed. If you want to merge now, I can add tests in another PR.

**Comment by BurntSushi** [comment]
Beautiful, thank you. This is really awesome.

**Comment by gereeter** [comment]
No problem - this library is fun to work with.

**Comment by emk** [comment]
@gereeter Wow, I _suspected_ that it was possible to do stuff like that, but you managed to implement it, and the code is amazingly elegant.

:clap:

**Comment by BurntSushi** [comment]
@emk Agreed! It is now on crates.io by the way. (I think we'll need a new `automaton` submodule for all of the new types. Don't want to clutter the top-level namespace too much.)

**Pull request merged**